ruby/ql/lib/codeql/ruby/frameworks/GraphQL.qll

```diff
@@ -0,0 +1,439 @@
+ * The user input data comes in through a GraphQL query, is parsed by the GraphQL
+ * library, and this method handles the request. Then the result is serialized
+ * into a GraphQL response on the way out.
+ *
+ * For example:
+ *
+ * ```rb
+ * module Mutation
+ *   class NameAnInstrument < BaseMutationn
+ *     argument :instrument_uuid, Types::Uuid,
+ *              required: true,
+ *              loads: ::Instrument,
+ *              as: :instrument,
+ *     argument :name, String, required: true
+ *
+ *     def load_instrument(uuid)
+ *       ::Instrument.find_by(uuid: uuid)
+ *     end
+ *
+ *     # GraphqlResolveMethod
+ *     def resolve(instrument:, name:)
+ *       instrument.set_name(name)
+ *     end
+ *   end
+ * end
+ * ```
+ */
+class GraphqlResolveMethod extends Method, HTTP::Server::RequestHandler::Range {
+  private GraphqlResolvableClass resolvableClass;
+
+  GraphqlResolveMethod() { this = resolvableClass.getMethod("resolve") }
+
+  override Parameter getARoutedParameter() { result = this.getAParameter() }
+
+  override string getFramework() { result = "GraphQL" }
+
+  /** Gets the mutation class containing this method. */
+  GraphqlResolvableClass getMutationClass() { result = resolvableClass }
+}
+
+/**
+ * A `load_*` method on a sub-class of either
+ * `GraphQL::Schema::RelayClassicMutation` or
+ * `GraphQL::Schema::Resolver`.
+ *
+ * This method takes user input (some kind of ID or specifier) and is intended
+ * to resolve the domain object using that ID.
+ *
+ * For example:
+ *
+ * ```rb
+ * module Mutation
+ *   class NameAnInstrument < BaseMutationn
+ *     argument :instrument_uuid, Types::Uuid,
+ *              required: true,
+ *              loads: ::Instrument,
+ *              as: :instrument,
+ *     argument :name, String, required: true
+ *
+ *     # GraphqlLoadMethod
+ *     def load_instrument(uuid)
+ *       ::Instrument.find_by(uuid: uuid)
+ *     end
+ *
+ *     def resolve(instrument:, name:)
+ *       instrument.set_name(name)
+ *     end
+ *   end
+ * end
+ * ```
+ */
+class GraphqlLoadMethod extends Method, HTTP::Server::RequestHandler::Range {
+  private GraphqlResolvableClass resolvableClass;
+
+  GraphqlLoadMethod() {
+    this.getEnclosingModule() = resolvableClass and
+    this.getName().regexpMatch("^load_.*")
+  }
+
+  override Parameter getARoutedParameter() { result = this.getAParameter() }
+
+  override string getFramework() { result = "GraphQL" }
+
+  /** Gets the mutation class containing this method. */
+  GraphqlResolvableClass getMutationClass() { result = resolvableClass }
+}
+
+/**
+ * A `ClassDeclaration` for a class that extends
+ * `GraphQL::Schema::Object`.
+ *
+ * These classes can contain `field` definitions for the graphql schema.
+ *
+ * For example:
+ *
+ * ```rb
+ * class Types::User < GraphQL::Schema::Object
+ *   field :email, String
+ * end
+ * ```
+ */
+class GraphqlFieldableClass extends ClassDeclaration {
```

Comment:
Is there a reason this class exists, separate from `GraphqlSchemaObjectClass`? They seem like fundamentally the same thing, except this has the extra member predicate `getAFieldDefinitionMethodCall`. If possible, it would be nice to merge the two together - I feel that `GraphqlSchemaObjectClass` is a better, more self-descriptive name.

Reply:
Yes, this class exists because it's where I would eventually like to add support for the [`GraphQL::Schema::Interface`](https://graphql-ruby.org/type_definitions/interfaces.html#definition-methods) functionality, where you can create reusable modules that contain a common set of fields. However, I'm not sure exactly how I'll implement that, and since those are ruby `module`s, not `class`es, It likely wouldn't fit in here anyway. Maybe I'd be able to use like a `union` type or something. Anyway, good call-out. I'll remove this for now.